tooling/greybox_fuzzer/src/mutation/mod.rs

```diff
@@ -0,0 +1,769 @@
+    /// If we hit any arrays along the way, we enable the structural mutations for selection. If in the end structural mutation is chosen, the probability at each level is the same
+    /// The result is the input value generated by the mutation and an optional directive to mutate an array structurally
+    pub fn mutate_input_value(
+        &self,
+        abi_type: &AbiType,
+        previous_input: &InputValue,
+        prng: &mut XorShiftRng,
+        weight_tree_node: &NodeWeight,
+        mutation_weight: u32,
+        arrays_hit: usize,
+    ) -> (InputValue, Option<usize>) {
+        // Check if this is a leaf type. If it is and the element is in an array,
+        // then we can trigger structural mutations for one of the arrays this element is in by randomly selecting the level
+        match abi_type {
+            AbiType::Boolean
+            | AbiType::Field
+            | AbiType::Integer { .. }
+            | AbiType::String { .. } => match BASIC_TOP_LEVEL_MUTATION_CONFIGURATION.select(prng) {
+                configurations::TopLevelMutation::Value => (),
+                configurations::TopLevelMutation::Structure => {
+                    if !arrays_hit.is_zero() {
+                        return (previous_input.clone(), Some(prng.gen_range(1..=arrays_hit)));
+                    }
+                }
+            },
+            AbiType::Array { .. } | AbiType::Struct { .. } | AbiType::Tuple { .. } => (),
+        }
+        match abi_type {
+            // Boolean only has 2 values, there is no point in performing complex logic
+            AbiType::Boolean => {
+                (InputValue::Field(FieldElement::from(prng.gen_range(0u32..=1u32))), None)
+            }
+            AbiType::Field => (
+                mutate_field_input_value(
+                    previous_input,
+                    self.full_dictionary.get_field_dictionary(),
+                    prng,
+                ),
+                None,
+            ),
+            AbiType::Integer { sign, width } => (
+                mutate_int_input_value(
+                    previous_input,
+                    sign,
+                    *width,
+                    self.full_dictionary.get_int_dictionary(),
+                    prng,
+                ),
+                None,
+            ),
+            AbiType::String { length: _ } => (
+                mutate_string_input_value(
+                    previous_input,
+                    prng,
+                    self.full_dictionary.get_int_dictionary(),
+                ),
+                None,
+            ),
+            AbiType::Array { length, typ } => {
+                let length = *length as usize;
+                let input_vector = match previous_input {
+                    InputValue::Vec(previous_input_vector) => previous_input_vector,
+                    _ => panic!("Mismatch of AbiType and InputValue should not happen"),
+                };
```

Comment:
nit:

```suggestion
                let InputValue::Vec(input_vector) = previous_input else {
                    panic!("Mismatch of AbiType and InputValue should not happen")
                };
```

Reply:
This can also be done in the `Struct` and `Tuple` cases.

Reply:
Thanks